README.md
Based on amazing Homespan library https://github.com/HomeSpan/HomeSpan by Gregg E. Berman
Implementing remote, sleeping devices that can last long on battery
So far only temperature, humidity and light, maybe more later


changelog.txt
2022-11-27:
- 0.4.6 - disable OTA on webserver as it only works inside LAN and takes a lot of memory 
2022-11-26:
- 0.4.5 - 
        - consolidate checking volts and checking charging in one task
        - adding DEBUG_XTASKS to estimate stack size for vTasks
        - optimizing stack size for led_blink() and check_volts
        - change xTaskCreatePinnedToCore to xTaskCreate - let software decide which core to use
        - add: board_build.partitions = min_spiffs.csv to platformio.ini to enable Inspector in PIO
- 0.4.4 - free heap checking in check_volts(), restart ESP if < MIN_ALLOWED_HEAP_SIZE (20kB)
- 0.4.3 - fw_version reworked again: "0.4.3" - only numbers here, major: 0-99, minor: 0-9, patch: 0-9, so 0.4.3 would be 43, no seconds in fw_version
- 0.4.2 - fw_version, serial_number: both changed on bridge, fw_version has date/time of compilation, serial_number has both macs
2022-11-25:
- 0.4.1 - FW update for BRIDGE finished
- 0.4.0 -
        - FW update from web on press "ID" in Eve for HomeKit (check how to in Home.app)
        - ESP32-S2 discontinued - not enough memory
        - new SN introduced: last 6 char of new MAC and last 6 char of old MAC
- 0.3.4 - bootcount implemented on SENSOR DEVICES
- 0.3.2 - charging showing on HomeKit, battery level on HomeKit
2022-11-23:
- 0.2.0 - added battery to sensors (in %)


hub/src/DEV_Identify.h
// NOT NEEDED - IN MAIN FILE




//////////////////////////////////
//   DEVICE-SPECIFIC SERVICES   //
//////////////////////////////////

// struct DEV_Identify : Service::AccessoryInformation 
// {

//   int nBlinks;                    // number of times to blink built-in LED in identify routine
//   SpanCharacteristic *identify;   // reference to the Identify Characteristic
  
//   DEV_Identify(const char *name, const char *manu, const char *sn, const char *model, const char *version, int nBlinks) : Service::AccessoryInformation()
//   {
    
//     new Characteristic::Name(name);                   // create all the required Characteristics with values set based on above arguments
//     new Characteristic::Manufacturer(manu);
//     new Characteristic::SerialNumber(sn);    
//     new Characteristic::Model(model);
//     new Characteristic::FirmwareRevision(version);
//     identify=new Characteristic::Identify();          // store a reference to the Identify Characteristic for use below

//     this->nBlinks=nBlinks;                            // store the number of times to blink the LED

//     // pinMode(homeSpan.getStatusPin(),OUTPUT);          // make sure LED is set for output
//   }

//   boolean update()
//   { 
//     #ifdef ERROR_RED_LED_GPIO    
//         LOG0("[%s]: hub called to identify, blinking ERROR LED...\n",__func__);
//         for(int i=0;i<nBlinks;i++)
//         {
//             digitalWrite(ERROR_RED_LED_GPIO,LOW);
//             delay(50);
//             digitalWrite(ERROR_RED_LED_GPIO,HIGH);
//             delay(50);
//         }
//     #elif defined(STATUS_LED_GPIO)
//         LOG0("[%s]: hub called to identify, blinking status LED...",__func__);
//         for(int i=0;i<nBlinks;i++)
//         {
//             digitalWrite(STATUS_LED_GPIO,LOW);
//             delay(50);
//             digitalWrite(STATUS_LED_GPIO,HIGH);
//             delay(50);
//         }
//     #else 
//         LOG0("[%s]: no LED defined - failed\n",__func__);
//         return false;
//     #endif
//     LOG0("[%s]: DONE\n",__func__);
//     return true;
//   }

// };


hub/src/devices_config.h

/*
ESP32-S2  is NOT OK, it has NOT ENOUGH memory to perform FW update (heap size < 64kB when HomeSpan is working)
ESP32S    is OK
ESP32-S3  is OK
ESP32-C3  not tested yet
*/



// #elif DEVICE_ID == 0                    // ESP32-S2 - not enough memory - don't use it - FW update fails - kept for GPIO and config example only
//     #define HOSTNAME                  "bridge1"
//     #define BOARD_TYPE                2     // 1 = ESP32-S, 2 = ESP32-S2, 3 = ESP32-S3, 4 = ESP32-C3
// ---------------------------------------------------------------------------------------------------
#if DEVICE_ID == 1                    // ESP32-S
    #define HOSTNAME                  "bridge1"
    #define BOARD_TYPE                1     // 1 = ESP32-S, 2 = ESP32-S2, 3 = ESP32-S3, 4 = ESP32-C3
    // bridge fake MAC
    uint8_t FixedMACAddress[] =       {0x1A, 0xFF, 0x01, 0x01, 0x01, 0x01};
    #define STATUS_LED_GPIO           33
    #define ERROR_RED_LED_GPIO        32
    #define CONTROL_PIN               23

    // with MAX17048
    #define CUSTOM_SDA_GPIO           21   // override default SDA pin
    #define CUSTOM_SCL_GPIO           22   // override default SCL pin
    #define USE_MAX17048              1   // use "0" to disable
    #define POWER_GPIO                25  // GREEN, STDB PIN6 on TP4056, LOW on CHARGED (LED ON),       comment out if not in use - don't use "0" here
    #define CHARGING_GPIO             26  // RED,   CHRG PIN7 on TP4056, LOW during CHARGING (LED ON),  comment out if not in use - don't use "0" here
    
  #pragma message "compilation for: DEVICE_ID=1, ESP32-S, " HOSTNAME
// ---------------------------------------------------------------------------------------------------

#else
  #error "Wrong DEVICE_ID chosen"
#endif


// assigning MODEL and checking if proper board is selected
#ifndef BOARD_TYPE
  #error BOARD_TYPE not defined
#else
  #if (BOARD_TYPE == 1) and (!defined(CONFIG_IDF_TARGET_ESP32))
    #error wrong board selected in Arduino - choose ESP32DEV
  #endif
  #if (BOARD_TYPE == 2) and (!defined(CONFIG_IDF_TARGET_ESP32S2))
    #error wrong board selected in Arduino - choose S2
  #endif
  #if (BOARD_TYPE == 3) and (!defined(CONFIG_IDF_TARGET_ESP32S3))
    #error wrong board selected in Arduino - choose S3
  #endif
  #if (BOARD_TYPE == 4) and (!defined(CONFIG_IDF_TARGET_ESP32C3))
    #error wrong board selected in Arduino - choose C3
  #endif

  #if (BOARD_TYPE == 1)
    #define MODEL "ESP32"
    // #define ESP32_BOOT_TIME           350      // since power ON to GPIO ON, ms
    // #define ESP32_TAIL_TIME           15      // since GPIO OFF to power OFF, ms
    // #define ESP32_BOOT_TIME_EXTRA     20     // +180ms from power ON or hard reset    #ifdef PRINT_COMPILER_MESSAGES
    #ifdef PRINT_COMPILER_MESSAGES
      #pragma message "chosen BOARD_TYPE = ESP32"
    #endif
  #elif (BOARD_TYPE == 2)
    #define MODEL "ESP32S2"
    // #define ESP32_BOOT_TIME           35      // since power ON to GPIO ON, ms
    // #define ESP32_TAIL_TIME           25      // since GPIO OFF to power OFF, ms
    // #define ESP32_BOOT_TIME_EXTRA     180     // +180ms from power ON or hard reset
    #ifdef PRINT_COMPILER_MESSAGES
      #pragma message "chosen BOARD_TYPE = ESP32S2"
    #endif
  #elif (BOARD_TYPE == 3)
    #define MODEL "ESP32S3"
    // #define ESP32_IS_CHEATING 0       // not tested yet
    #ifdef PRINT_COMPILER_MESSAGES
      #pragma message "chosen BOARD_TYPE = ESP32S3"
    #endif
  #elif (BOARD_TYPE == 4)
    #define MODEL "ESP32C3"
    // #define ESP32_BOOT_TIME           30      // since power ON to GPIO ON, ms
    // #define ESP32_TAIL_TIME           10      // since GPIO OFF to power OFF, ms
    // #define ESP32_BOOT_TIME_EXTRA     155     // +150ms from power ON or hard reset
    #ifdef PRINT_COMPILER_MESSAGES
      #pragma message "chosen BOARD_TYPE = ESP32C3"
    #endif
  #else
    #error BOARD_TYPE not defined
  #endif
#endif
// assigning MODEL and checking if proper board is selected END